crates/bevy_input/src/keyboard.rs

```diff
@@ -43,19 +43,61 @@ pub struct KeyboardInput {
     pub window: Entity,
 }
 
-/// Updates the [`Input<KeyCode>`] resource with the latest [`KeyboardInput`] events.
+/// The stored logical key code for `Res<Input<KeyLogic>>`
+///
+/// We need to store the corresponding `KeyCode` because when modifiers are applied,
+/// the released logical key might be different from the pressed.
+/// For example, the sequence of physical keys:
+/// - Press shift
+/// - Press 'a'
+/// - Release shift
+/// - Release 'a',
+///
+/// Leads to the following underlying logical events:
+/// - Press shift
+/// - Press logic character key 'A'
+/// - Release shift
+/// - Release logic character key 'a'
+/// Note the different capitalization.
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Reflect)]
+#[cfg_attr(
+    feature = "serialize",
+    derive(serde::Serialize, serde::Deserialize),
+    reflect(Serialize, Deserialize)
+)]
+pub enum KeyLogic {
```

Comment:
Unifying these into a single type seems incorrect: it seems likely to increase confusion and boilerplate. See comment about about the HashMap.